### PR TITLE
Decline and Cancel reasons report update

### PR DESCRIPTION
**Pull request by eanders**
This is fed by https://github.com/greenriver/boston-cas/pull/237.
Detail for the request are here: https://www.pivotaltracker.com/story/show/168038272

### CODE REVIEW

File: app/models/warehouse_report/cas_declines.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      reason = row.decline_reason || row.administrative_cancel_reason
+      reason.squish.gsub(/Other.*/,'Other').strip
+    end.each_with_object(Hash.new(0)) do |reason, counts|
+      counts[reason] += 1
+    end.sort_by(&:last).reverse
+  end
+
+  def declines
+    @declines ||= report_source.declined.
+      started_between(start_date: start_date, end_date: end_date + 1.day)
```
Comment (awisema) [inline app/models/warehouse_report/cas_declines.rb]: Why `+ 1.day`?

File: app/controllers/warehouse_reports/cas/decline_reason_controller.rb [hunk 2]
```diff
@@ -10,18 +10,20 @@ class DeclineReasonController < ApplicationController
...
-      end.each_with_object(Hash.new(0)) { |reason,counts| counts[reason] += 1 }
-
-      @all_steps = report_source.
-        where(match_id: @matches.select(:match_id)).
-        order(decision_order: :asc).
-        group_by(&:match_id)
-
-      @reasons.sort_by(&:last).reverse
+      @report = WarehouseReport::CasDeclines.new(start_date: @range.start, end_date: @range.end)
+
```
Comment (awisema) [inline app/controllers/warehouse_reports/cas/decline_reason_controller.rb]: Clean up removed code?

File: app/views/warehouse_reports/cas/decline_reason/index.haml [hunk 3]
```diff
@@ -16,16 +16,29 @@
...
+            - @report.referrals_by_agency.each do |(program_name, sub_program_name), referral_count|
               %tr
-                %td= reason
-                %td= n
+                %td= program_name
+                %td= sub_program_name
+                %td= referral_count
+                %td
+                  - count = @report.declines_by_agency[[program_name, sub_program_name]] || 0
+                  - pecent = ((count / referral_count.to_f).round(2) * 100).round rescue 0
```
Comment (awisema) [inline app/views/warehouse_reports/cas/decline_reason/index.haml]: Typo: `percent`

File: app/models/warehouse_report/cas_declines.rb (hunk 1 above)
Comment (eanders) [reply app/models/warehouse_report/cas_declines.rb]: I believe it was to include items started on that day... though it was just a copy/paste from the previous version of the report.

**Pull request merged**